all: Undo custom exit codes

BenWiederhake opened the pull request:
This PR:
- fixes the exit codes for three utils; the bulk of utils was already fixed back in #3922
- fixes a weird difference in behavior of `tee -h` and `tee --help`
- reduces our customization of the GNU test `misc/usage_vs_getopt`

All in all, we should now pass `misc/usage_vs_getopt`. Let's see whether CI agrees!

Ping @tertsdiepraam, you touched on this topic a long time ago. What do you think? :)

BenWiederhake commented:
Hooray! It works. It also reveals yet another inconsistency in GNU ls:
```console
$ ls --invalid
ls: unrecognized option '--invalid'
Try 'ls --help' for more information.
[$? = 2]
$ ls --classify=invalid
ls: invalid argument 'invalid' for '--classify'
Valid arguments are:
  - 'always', 'yes', 'force'
  - 'never', 'no', 'none'
  - 'auto', 'tty', 'if-tty'
Try 'ls --help' for more information.
[$? = 1]
```

BenWiederhake commented:
Changes since last push:
- rebased on current main
- fix exit code for `--classify=invalid` and all other stringly-typed enum-like arguments
- fix even the weirdly behaving `--time-style`

BenWiederhake commented:
Changes since last push:
- Rebased on current main
- Moved some commits into #6173, which was merged separately.

Locally this passes `misc/usage_vs_getopt` again. Let's see whether CI agrees this time! :)

tertsdiepraam commented on src/uu/tee/src/tee.rs:
```diff
@@ -89,6 +89,16 @@ pub fn uu_app() -> Command {
         .override_usage(format_usage(USAGE))
         .after_help(AFTER_HELP)
         .infer_long_args(true)
+        // Since we use value-specific help texts for "--output-error", clap's "short help" and "long help" differ.
```
I'm not sure about this change. It's more compatible, sure, but is it better? I want to limit all sorts of edge cases in the code and might prefer disabling that test here. Or can we force the bit for `--output-error` to always be shown?

Note that this will be fixed by uutils-args too by the way, so maybe it just doesn't matter much either way.

BenWiederhake replied:
I don't quite understand? All that this piece of code does is force the bit for `--output-error` to always be shown, just like you suggest.

BenWiederhake commented:
Fancy new flake discovered in Windows CI:
```
Error: HttpError: request to https://api.github.com/repos/mozilla/sccache/releases/latest failed, reason: read ECONNRESET
Error: request to https://api.github.com/repos/mozilla/sccache/releases/latest failed, reason: read ECONNRESET
```
I guess we're (I am) making too many calls and we're getting rate-limited?

BenWiederhake commented:
Changes since last push: Nothing, just a rebase to show that it still works.

@tertsdiepraam Since the code already does what you suggest (I probably misunderstand you, please clarify), I didn't change anything :)

tertsdiepraam commented:
I kwant that I wanted it by including the same text in the short help (if possible), instead of forcing the long help for the whole util. Again though, not important enough to block on.

BenWiederhake commented:
Changes since last push:
- Rebased to current main, to show that it still works

Is there anything else needed?

sylvestre commented:
well done
